shell/platform/android/io/flutter/embedding/android/FlutterImageView.java

```diff
@@ -188,26 +169,23 @@ public boolean acquireLatestImage() {
     if (!isAttachedToFlutterRenderer) {
       return false;
     }
-    // There's no guarantee that the image will be closed before the next call to
-    // `acquireLatestImage()`. For example, the device may not produce new frames if
-    // it's in sleep mode, so the calls to `invalidate()` will be queued up
-    // until the device produces a new frame.
+    // 1. `acquireLatestImage()` may return null if no new image is available.
     //
-    // While the engine will also stop producing frames, there is a race condition.
+    // 2. There's no guarantee that the `onDraw()` called after `invalidate()`.
+    // For example, the device may not produce new frames if
+    // it's in sleep mode or some special Android devices so the calls to `invalidate()` will be
+    // queued up
+    // until the device produces a new frame.
     //
-    // To avoid exceptions, check if a new image can be acquired.
-    int imageOpenedCount = imageQueue.size();
-    if (currentImage != null) {
-      imageOpenedCount++;
-    }
-    if (imageOpenedCount < imageReader.getMaxImages()) {
-      final Image image = imageReader.acquireLatestImage();
```

Comment:
I feel like there is some misunderstanding from my side. The error message " Unable to acquire a buffer item, very likely client tried to acquire more than maxImages buffers" suggests that `imageReader.acquireLatestImage()` was called even though  `imageOpenedCount` is less than ` imageReader.getMaxImages()`.  Unless there's a bug in this logic, shouldn't it be preventing the `acquireLatestImage()` call?

Reply:
@blasten I guess maybe the newest available image will be also counted by the code in ImageReader.cpp.
Let me have a deeper view in ImageReader.cpp

Reply:
@blasten I read the code and document of `acquireLatestImage()` in ImageReader.java carefully and finally knows why the error message comes even though the openedCount is less than maxImages.

The code:
https://cs.android.com/android/platform/superproject/+/master:frameworks/base/media/java/android/media/ImageReader.java;drc=master;l=410?q=two

The doc of `acquireLatestImage()` says:
```
     * Note that {@link #getMaxImages maxImages} should be at least 2 for
     * {@link #acquireLatestImage} to be any different than {@link #acquireNextImage} -
     * discarding all-but-the-newest {@link Image} requires temporarily acquiring two
     * {@link Image Images} at once. Or more generally, calling {@link #acquireLatestImage}
     * with less than two images of margin, that is
     * {@code (maxImages - currentAcquiredImages < 2)} will not discard as expected.
     * </p>
```
So it need at least **two reserved slots** in buffer before one calling of `acquireLatestImage()`. And the implementation code is as same as it described:

```java
    public Image acquireLatestImage() {
        Image image = acquireNextImage();
        if (image == null) {
            return null;
        }
        try {
            for (;;) {
                Image next = acquireNextImageNoThrowISE();
                if (next == null) {
                    Image result = image;
                    image = null;
                    return result;
                }
                image.close();
                image = next;
            }
        } finally {
            if (image != null) {
                image.close();
            }
        }
    }

```

Every `nativeImageSetup()` will need a new BufferItem in cpp,  `acquireNextSurfaceImage()`  calls  `nativeImageSetup()`. then looking at the `acquireLatestImage()` logic, it calls `acquireNextImage()` and calls `acquireNextImageNoThrowISE` in the for-loop and close the last image unless the next is null. So I can really need towo slots.

The code of this pull request fixs onDraw() may be lesser then invalidate(), and try to make the code logic neat and easy to maintain by removing the queue and the counter.